src/NuGet.Core/NuGet.Packaging/Rules/DependenciesGroupsForEachTFMRule.cs

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Xml.Linq;
+using NuGet.Client;
+using NuGet.Common;
+using NuGet.ContentModel;
+using NuGet.Frameworks;
+using NuGet.Packaging.Core;
+using NuGet.RuntimeModel;
+
+namespace NuGet.Packaging.Rules
+{
+    internal class DependenciesGroupsForEachTFMRule
+        : IPackageRule
+    {
+        private static readonly NuGetFramework Net00 = NuGetFramework.Parse("net00");
+
+        public string MessageFormat => AnalysisResources.DependenciesGroupsForEachTFMHasNoExactMatch;
+
+        private string CompatMatchFoundWarningMessageFormat => AnalysisResources.DependenciesGroupsForEachTFMHasCompatMatch;
+
+        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)
+        {
+            var files = builder.GetFiles().
+                Where(t => PathUtility.GetPathWithDirectorySeparator(t).Count(m => m == Path.DirectorySeparatorChar) > 1);
+            var packageNuspec = ExtractTFMsFromNuspec(builder.GetNuspec());
+            var (compatNotExactMatches, noExactMatchesFromFile, noExactMatchesFromNuspec) = Categorize(files, packageNuspec);
+            return GenerateWarnings(compatNotExactMatches, noExactMatchesFromFile, noExactMatchesFromNuspec);
+        }
+
+        internal IEnumerable<PackagingLogMessage> GenerateWarnings(HashSet<NuGetFramework> compatNotExactMatches, HashSet<NuGetFramework> noExactMatchesFromFile, HashSet<NuGetFramework> noExactMatchesFromNuspec)
+        {
+            (string noExactMatchString, string compatMatchString) = GenerateWarningString(noExactMatchesFromFile, noExactMatchesFromNuspec, compatNotExactMatches);
+
+            var issues = new List<PackagingLogMessage>();
+
+            if (noExactMatchesFromFile.Count != 0 || noExactMatchesFromNuspec.Count != 0)
+            {
+                issues.Add(PackagingLogMessage.CreateWarning(string.Join("\n", MessageFormat, noExactMatchString),
```

Comment:
Move the functionality of the `String.Join` to `GenerateWarningString`, where you can use the string builder without creating temporary strings.